### PR TITLE
test(calling): voicemail module test cases added

### PR DESCRIPTION
**Pull request by meta-dipanshu-sharma**
# COMPLETES #< [SPARK-422971](https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-422971) >

## This pull request addresses

Code coverage for voicemail module files were less.  Need to write UT to cover the uncovered code.

## by making the following changes

Added UTs for following files:
1. BroadworksBackendConnector.test.ts
2. UcmBackendConnector.test.ts
3. Voicemail.test.ts

Some code refactor in below files:
1. BroadworksBackendConnector.ts
2. UcmBackendConnector.ts
3. Utils.ts

CODE COVERAGE REPORT:
PREVIOUSLY
Screenshot 2023-09-24 at 11 40 01 PM

CURRENT:
Screenshot 2023-09-26 at 2 21 30 PM

### Change Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [x] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

### CODE REVIEW

**Review by karthikraji2020 — COMMENTED** [review]
Left few comments. please check.

File: packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts [hunk 1]
```diff
@@ -60,6 +76,70 @@ describe('Voicemail Broadworks Backend Connector Test case', () => {
...
+          file: 'BroadworksBackendConnector',
+          method: 'voicemailMarkAsUnread',
+        }
+      );
+    });
+
+    it('verify failure case for the delete voicemail case when response is not ok', async () => {
+      const response = await broadworksBackendConnector.deleteVoicemail('dummy');
+
+      expect(response.message).toBe('FAILURE');
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts]: Done

File: packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts [hunk 2]
```diff
@@ -108,6 +188,14 @@ describe('Voicemail Broadworks Backend Connector Test case', () => {
       );
     });
 
+    it('verify no response case when token have invalid userid', async () => {
+      broadworksBackendConnector['bwtoken'] = 'bwtoken.eyJhbGciOiJIUzI1NiJ9';
+      const response = await broadworksBackendConnector.init();
+
+      /* eslint-disable no-unused-expressions */
+      expect(response).toBeUndefined;
```
[the remark this answers is not in the document]
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts]: Done

File: packages/calling/src/Voicemail/UcmBackendConnector.ts [hunk 3]
```diff
@@ -254,10 +260,11 @@ export class UcmBackendConnector implements IUcmBackendConnector {
     });
     const contentInfo = response?.body as UcmVMContentResponse;
     const respHeaders = response.headers;
+    const statusCode = response.statusCode;
     const mediaType = respHeaders?.mediatype as string;
     const mediaContent = contentInfo as string;
     const responseDetails = {
-      statusCode: response.statusCode as number,
+      statusCode: statusCode as number,
```
Comment (BhargavSatya) [inline packages/calling/src/Voicemail/UcmBackendConnector.ts:267]: ` as number` has known to be problematic and we are changing it whenever we encounter it during new changes.
```suggestion
      statusCode: Number(statusCode),
```

File: packages/calling/src/Voicemail/BroadworksBackendConnector.ts [hunk 4]
```diff
@@ -465,7 +465,9 @@ export class BroadworksBackendConnector implements IBroadworksCallBackendConnect
       return responseDetails;
     } catch (err: unknown) {
       /* Catch the exception error code from try block, return the error object to user */
-      const errorInfo = err as WebexRequestPayload;
+      const errorInfo = {
+        statusCode: err instanceof Error ? Number(err.message) : '',
```
Comment (BhargavSatya) [inline packages/calling/src/Voicemail/BroadworksBackendConnector.ts:469]: Can we not set '' as default statusCode. Put 422 or equivalent code.

```suggestion
        statusCode: err instanceof Error ? Number(err.message) : 422,
```

**Review by BhargavSatya — CHANGES_REQUESTED** [review]
Minor comments. 
Can we try to increase branch coverage to minimum of 85% as well?

File: packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts [hunk 5]
```diff
@@ -306,8 +413,69 @@ describe('Voicemail Broadworks Backend Connector Test case', () => {
...
+
+      const response = await broadworksBackendConnector.getVoicemailList(0, 20, SORT.DESC, true);
+
+      const voicemailResponseInfo = {
+        voicemailList: [{}],
+      };
+
+      const responseDetails = {
+        statusCode: NO_VOICEMAIL_STATUS_CODE,
+        data: voicemailResponseInfo,
```
Comment (Shreyas281299) [inline packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts:462]: In the above test an empty voicemailResponseInfo is initialised to [] and here we have [{}]. Arent they the same array?

File: packages/calling/src/Voicemail/UcmBackendConnector.test.ts [hunk 6]
```diff
@@ -64,6 +88,31 @@ describe('Voicemail UCM Backend Connector Test case', () => {
     });
   });
 
+  it('verify successful voicemailContent', async () => {
+    const voiceMailPayload = <VoicemailResponseEvent>(<unknown>mockVoicemailContentResponse);
+
+    webex.request.mockResolvedValueOnce(voiceMailPayload);
+    const response = await (await ucmBackendConnector.getVoicemailContent(messageId)).data;
```
Comment (Shreyas281299) [inline packages/calling/src/Voicemail/UcmBackendConnector.test.ts:95]: Do we require two await here?

File: packages/calling/src/Voicemail/WxCallBackendConnector.test.ts [hunk 7]
```diff
@@ -662,6 +680,60 @@ describe('Voicemail webex call Backend Connector Test case', () => {
...
+      });
+    });
+
+    it('verify empty voicemail list data when response data is in invalid format', async () => {
+      webex.request.mockResolvedValueOnce(getInvalidVoicemailListJsonWxC);
+
+      const response = await wxCallBackendConnector.getVoicemailList(0, 20, SORT.DESC, true);
+
+      const voicemailResponseInfo = {
+        voicemailList: [{}],
```
Comment (Shreyas281299) [inline packages/calling/src/Voicemail/WxCallBackendConnector.test.ts:716]: Same comment as above.

**Review by Shreyas281299 — COMMENTED** [review]
Minor comments.

File: packages/calling/src/Voicemail/UcmBackendConnector.ts (hunk 3 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/UcmBackendConnector.ts]: Done

File: packages/calling/src/Voicemail/BroadworksBackendConnector.ts (hunk 4 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/BroadworksBackendConnector.ts:469]: We're setting default statusCode as 422 in `serviceErrorCodeHandler` in the default switch statement when it doesn't matches with any of the status code. If we want to set it default here then we have to update it at multiple places wherever it is used. Do you still suggest changing this at multiple places or current implementation is acceptable?

File: packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts (hunk 5 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/BroadworksBackendConnector.test.ts:462]: No, they are testing different conditions:
1. Empty array is testing this condition: https://github.com/webex/webex-js-sdk/blob/next/packages/calling/src/Voicemail/BroadworksBackendConnector.ts#L235

2. [{}] is testing this: https://github.com/webex/webex-js-sdk/blob/next/packages/calling/src/Voicemail/BroadworksBackendConnector.ts#L240

File: packages/calling/src/Voicemail/UcmBackendConnector.test.ts (hunk 6 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/UcmBackendConnector.test.ts]: Not required, updated the code. Thanks for noticing!

File: packages/calling/src/Voicemail/WxCallBackendConnector.test.ts (hunk 7 above)
Comment (meta-dipanshu-sharma) [reply packages/calling/src/Voicemail/WxCallBackendConnector.test.ts:716]: replied above.

File: packages/calling/src/Voicemail/BroadworksBackendConnector.ts (hunk 4 above)
Comment (BhargavSatya) [reply packages/calling/src/Voicemail/BroadworksBackendConnector.ts:478]: Thats fine . Ignore this if it was like this

**Comment by meta-dipanshu-sharma** [comment]
> Minor comments. Can we try to increase branch coverage to minimum of 85% as well?

@BhargavSatya I have made BroadworksBackendConnector file's branch coverage to >85%. For UcmBackendConnector file, we were not able to cover some part of code and those branches are part of those lines only so could not cover them as well.

**Review by Shreyas281299 — APPROVED** [review]

**Review by BhargavSatya — APPROVED** [review]

**Pull request merged**